### ISSUE TITLE
How global is : global?

### ISSUE THREAD

**Issue by perpen** [flosell/lambdacd#162]
Looking at the code for control-flow/junction it seems like a global value set by a condition step would not be transmitted to the success or failure steps.
So global means "global to the current execution branch" rather than "global to the pipeline". Not that useful.
My potential use case here is `(junction deploy process-success process-failure)` where the`process-` functions use the `:out` from `deploy`

**Comment by flosell** [comment]
Global means "passed to all subsequent steps in the pipeline". 

So you are correct, `junction` does this wrong: Since `junction` always executes the condition first, then one of the two branches, it's reasonable to assume that junction behaves as if steps were sequential, passing the results of the condition to the branch-steps the way [serial-step-result-producer](https://github.com/flosell/lambdacd/blob/8f58bb8465a8baa023a88373249e96342901b6bc/src/clj/lambdacd/execution/internal/serial_step_result_producer.clj) does it for `run` and others.

The only thing I'm worried about is that this would be a breaking change: 
People most likely built pipelines depending on the current behavior (e.g. by depending on the fact that the branch-step will get the result of the step in front of the junction) which is why I'd like to try to find a clean, non-breaking migration path for this fix.

Just out of curiosity: What are you trying to achieve with this junction? What is it the process-success and process-failure steps are doing?

**Comment by perpen** [comment]
Example: the failure step could create an alert to our monitoring system, the alert message being the output of the deploy step.
Re. breaking change: Better to create a new step type then, but I cannot think of an expressive name. `branch`? `junction!`?

**Comment by flosell** [comment]
Ok, finally got around to looking at this in more detail. Turns out, fixing this didn't even involve a breaking change: What `run` does is [pass the previous steps results as well as the arguments of the parent step](https://github.com/flosell/lambdacd/blob/dacea7bc8e0d6ece928e1c626a7f17559435ca35/test/clj/lambdacd/steps/control_flow_test.clj#L203-L211) so for `junction` we just needed to add the condition-steps results.

In short, the success and failure steps now [receive the results of the condition step as well as the arguments passed into the junction](https://github.com/flosell/lambdacd/blob/0f5ac3af7d8e87b90f0374b349646825ad04ab72/test/clj/lambdacd/steps/control_flow_test.clj#L313-L336).
